src/Halogen/Components/Select/Rotary/Component.purs

```diff
@@ -0,0 +1,348 @@
+
+import Control.Monad (when)
+import Control.Monad.Aff (Aff())
+import Control.Monad.Aff.AVar (AVAR())
+import Control.Monad.Eff.Random (RANDOM())
+import Control.Monad.Maybe.Trans as Mt
+import Control.Coroutine.Aff (produce)
+import Control.Coroutine.Stalling as SCR
+
+import Data.Array as Arr
+import Data.Either as E
+import Data.Date (Now())
+import Data.Functor (($>))
+import Data.Functor.Eff (liftEff)
+import Data.Lens ((.~), (?~))
+import Data.Maybe as M
+import Data.Foldable as F
+import Data.StrMap as Sm
+import Data.Int as Int
+import Data.NonEmpty as Ne
+import Data.ExistsR as Er
+
+import CSS.Display (display, inlineBlock)
+import CSS.Display (position, relative)
+import CSS.Geometry (width, left, marginLeft)
+import CSS.Overflow (overflow, hidden)
+import CSS.Selector
+  (Selector(), Predicate(AttrVal), Refinement(..), (**), (##))
+import CSS.Size (px)
+import CSS.String (fromString)
+import CSS.Stylesheet (CSS(), (?))
+import CSS.TextAlign (textAlign, center)
+
+import Halogen hiding (Prop())
+import Halogen.HTML.Core (Prop(..), attrName, className, ClassName())
+import Halogen.HTML.Indexed as H
+import Halogen.HTML.Events.Indexed as E
+import Halogen.HTML.Events.Handler as E
+import Halogen.HTML.Properties.Indexed as P
+import Halogen.HTML.CSS as CSS
+import Halogen.HTML.Properties.Indexed (IProp())
+import Halogen.Query.EventSource (EventSource(..))
+
+import DOM (DOM())
+import DOM.HTML (window)
+import DOM.HTML.Window as Win
+import DOM.HTML.Types as Ht
+import DOM.Event.EventTarget as Etr
+import DOM.Event.EventTypes as Etp
+
+import Halogen.Components.Select.Utils.Random (genKey)
+import Halogen.Components.Select.Utils.DOM
+  (getComputedStyle, getClientRects, getScreen)
+import Halogen.Components.Select.Utils.Array (repeat, shift)
+import Halogen.Components.Select.Utils.NonEmpty (liftNonEmpty)
+
+import Halogen.Components.Select.Rotary.Component.State
+import Halogen.Components.Select.Rotary.Component.Query
+
+type AffSel e =
+  Aff ( dom :: DOM
+      , random :: RANDOM
+      , now :: Now
+      , avar :: AVAR |e)
+type RotarySelectorDSL e = ComponentDSL State Query (AffSel e)
+
+wrapperClass :: ClassName
+wrapperClass = className "rotary-selector-wrapper"
+
+draggedClass :: ClassName
+draggedClass = className "rotary-selector-dragged"
+
+itemClass :: ClassName
+itemClass = className "rotary-selector-item"
+
+dataRotaryKey :: forall i r. String -> IProp r i
+dataRotaryKey = Unsafe.Coerce.unsafeCoerce nonIndexed
+  where
+  nonIndexed :: String -> Prop i
+  nonIndexed = Attr M.Nothing (attrName "data-rotarykey")
+
+draggableScreens :: Int
+draggableScreens = 3
+
+type RotarySelectorConfig r =
+  {
+    itemRender :: M.Maybe (Option r -> ComponentHTML Query)
+  , itemWidth :: Number
+  , visibleItemCount :: M.Maybe Number
+  , items :: Ne.NonEmpty Array (Option r)
+  }
+
+comp
+  :: forall r e
+   . RotarySelectorConfig r
+  -> { component :: Component State Query (AffSel e)
+     , unpack :: OptionR -> Option r
```

Comment:
I think any time we have some arbitrary way of unpacking an exists we'll have this issue. I went through quite a few different approaches trying something similar when working on the notebook, and in the end couldn't make it work entirely safely, and instead ended up with to/from prisms from a sum type. That can still be made to fail if you really try (when the prism returns a Nothing it turns into a `Halt` in `HalogenDSL`), but at least that way it fails in a way we can observe rather than it "breaking" the type system.

Maybe @jonsterling would have another suggestion, as he helped me out with a lot of that stuff... but I suspect we do have to put the `r` type in `Query`, etc. to make it type safe.